plugins/gradle/tooling-extension-impl/src/org/jetbrains/plugins/gradle/tooling/tasks/DependenciesReport.groovy

```diff
@@ -54,191 +54,146 @@ class DependenciesReport extends DefaultTask {
+                                         ProjectNameFunction projectNameFunction) {
 
-    Set<RenderableDependency> getChildren() {
-      Set<RenderableDependency> out = new LinkedHashSet()
+      def selectedOrAttempted = dependency instanceof ResolvedDependencyResult
+        ? dependency.selected.id
+        : (dependency as UnresolvedDependencyResult).attempted
+      long id = idGenerator.getId(selectedOrAttempted)
+      DependencyNode alreadySeenNode = added.get(id)
+      if (alreadySeenNode != null) {
+        return new ReferenceNode(id)
+      }
 
-      for (def dependencyResult in this.module.getDependencies()) {
-        if (dependencyResult instanceof UnresolvedDependencyResult) {
-          out.add(new RenderableUnresolvedDependencyResult((UnresolvedDependencyResult)dependencyResult))
+      AbstractDependencyNode node
+      if (dependency instanceof ResolvedDependencyResult) {
+        def componentId = dependency.selected.id
+        if (componentId instanceof ProjectComponentIdentifier) {
+          node = new ProjectDependencyNodeImpl(id, projectNameFunction.fun(componentId))
+        }
+        else if (componentId instanceof ModuleComponentIdentifier) {
+          node = new ArtifactDependencyNodeImpl(id, componentId.getGroup(), componentId.getModule(), componentId.getVersion())
         }
         else {
-          out.add(new CustomRenderableDependencyResult((ResolvedDependencyResult)dependencyResult))
+          node = new UnknownDependencyNode(id, componentId.displayName)
+        }
+        node.resolutionState = RESOLVED
+        if (!dependency.selected.selectionReason.descriptions.isEmpty()) {
+          node.selectionReason = dependency.selected.selectionReason.descriptions.last().description
         }
+        added.put(id, node)
+        def iterator = dependency.selected.dependencies.iterator()
+        while (iterator.hasNext()) {
+          DependencyResult child = iterator.next()
+          node.dependencies.add(toNode(child, added, idGenerator, projectNameFunction))
+        }
+      } else if (dependency instanceof UnresolvedDependencyResult) {
+        node = new UnknownDependencyNode(id, dependency.attempted.displayName)
+        node.resolutionState = UNRESOLVED
+        added.put(id, node)
+      } else {
+        throw new IllegalStateException("Unsupported dependency result type: $dependency")
       }
 
-      return out
-    }
-  }
-
-  private static class CustomRenderableDependencyResult extends AbstractRenderableDependencyResult {
-    private final ResolvedDependencyResult dependency
-
-    CustomRenderableDependencyResult(ResolvedDependencyResult dependency) {
-      this.dependency = dependency
+      return node
     }
 
-    ResolutionState getResolutionState() {
-      return this.dependency.isConstraint() ? ResolutionState.RESOLVED_CONSTRAINT : ResolutionState.RESOLVED
-    }
+    private static class IdGenerator {
+      private THashMap<Object, Long> idMap = new THashMap<>()
```

Comment:
Using THashMap here since we still support Java 6, but I wanted an efficient putIfAbsent method.